app/utils/queryBuilder.ts

```diff
@@ -0,0 +1,378 @@
+import { useMemo } from "react";
+import { gql } from "graphql-request";
+
+// todo: test quote escape
+
+const BASE_QUERY_STRING = `
+    query $queryName {
+        $tableName(limit: $limit, offset: $offset, order_by: $orderBy where: $where) {
+            $columns
+        }
+    }`;
+
+/**
+ * The types we currently support as filter values
+ *
+ */
+type FilterValue = string | number | boolean | number[];
+
+/**
+ * Interface for a single filter that can be
+ * converted into a graphql `where` expression
+ */
+export interface Filter {
+  /**
+   * Arbitrary but must uniquely identify the filter by name amongst all
+   * other filters in the same group
+   */
+  id: string;
+  /**
+   * Hasura comparison operator, e.g. _eq, _gte
+   */
+  operator:
+    | "_gte"
+    | "_lte"
+    | "_gt"
+    | "_eq"
+    | "_neq"
+    | "_is_null"
+    | "_ilike"
+    | "_in"
+    | "_nin";
```

Comment:
hasura supports more operators, but i've only added those that we're actually using. we can add more as needed 👍